### PR TITLE
Add mount option to disable additional checksum for S3 Outposts write support

### PR DESCRIPTION
**Pull request by sauraank**
## Description of change
Added the disabling of checksums based on signing name as "s3-outposts".
Commented out the mount option of Disabling additional checksums to support write in Outposts since the decision is not yet made. Users of Outposts will have to pass `--disable-checksum` to support write, if we go this path.
This is draft PR, since we have not finalised our decision on the way to do it. But, whatever decision we take we will need to make similar changes in `upload.rs` module. 

Passed the trailing checksum boolean to S3FileSystemConfig and then to uploader. Using this config, set the trailing checksum as false  and bypassing the checksum verification for the case above.
Manually tested it works with write on Outposts. I will need to add tests.

Relevant issues: 
#516 

## Does this change impact existing behavior?
Yes, it disables additional checksums in case of writing on Outpost bucket (in this way, set by user). 

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/upload.rs [hunk 1]
```diff
@@ -64,6 +69,7 @@ pub struct UploadRequest<Client: ObjectClient> {
     key: String,
     next_request_offset: u64,
     hasher: Hasher,
+    set_checksum: bool,
```
Comment (passaro) [inline mountpoint-s3/src/upload.rs:72]: Use consistent name here

File: mountpoint-s3/src/upload.rs [hunk 2]
```diff
@@ -88,6 +94,7 @@ impl<Client: ObjectClient> UploadRequest<Client> {
             key: key.to_owned(),
             next_request_offset: 0,
             hasher: Hasher::new(),
```
Comment (passaro) [inline mountpoint-s3/src/upload.rs:96]: The `trailing_checksums` flag should also affect hashing: if disabled, we don't need the `Hasher`, or at least we don't need to calculate the checksum in `write`.

File: mountpoint-s3/src/upload.rs [hunk 3]
```diff
@@ -125,7 +132,7 @@ impl<Client: ObjectClient> UploadRequest<Client> {
         let size = self.size();
         let checksum = self.hasher.finalize();
         self.request
-            .review_and_complete(move |review| verify_checksums(review, size, checksum))
+            .review_and_complete(move |review| verify_checksums(review, size, checksum, self.set_checksum))
```
Comment (passaro) [inline mountpoint-s3/src/upload.rs:135]: You could skip the review completely and just call `PutObjectRequest::complete()` here.

Comment (passaro) [reply mountpoint-s3/src/upload.rs:135]: Alternatively, you could still use a review to verify just the size. In either cases, you do not want to calculate the checksum if not using it.

**Review by passaro — CHANGES_REQUESTED** [review]
While we discuss whether this is the right approach, I've commented on a couple of implementation details.

**Comment by chrsmia** [comment]
I believe this also will be required for support with s3-snow compatible storage on Snowball Edge.  So there's greater need than just for Outposts.
SHA checksums are not matching causing write activities to fail.

**Comment by jamesbornholt** [comment]
Moved to #849.

**Pull request closed without merging**